tests/test_authz_authorizer.py

```diff
@@ -38,37 +32,161 @@ def test__init__(self):
 
-    def test__repr__(self):
+    def test__repr__(self) -> None:
         assert repr(self.authz) == (
             "Authorizer(auth_jwt=<jwt>, "
             "resource_name='test_resource', "
             "permission_name='permission_name')"
         )
 
-    def test__set_policy_w_scope(self):
+    def test_validate_one_with_expired(self) -> None:
+        expiration_negative = int((datetime.utcnow() - timedelta(seconds=1)).timestamp())
+        with pytest.raises(Unauthorized):
+            self._make_authorizer(
+                {
+                    **self.token_payload,
+                    "allow": {"test_resource": {"permission_name": {"allow": ALL}}},
+                    "exp": expiration_negative,
+                }
+            )
+
+
+# pylint: disable=too-many-public-methods
+class TestAuthorizerSetupMethod:
+    @staticmethod
+    def _mocked_make_authorizer(token_payload: dict) -> Authorizer:
+        with patch("lbz.authz.authorizer.decode_jwt", lambda _: token_payload):
+            return Authorizer("xx", "test_resource", "permission_name")
+
+    @pytest.fixture(autouse=True)
+    def setup_method(self, base_auth_payload) -> None:
+        # pylint: disable=attribute-defined-outside-init
+        self.token_payload = base_auth_payload
+        self.authz = self._mocked_make_authorizer(self.token_payload)
+
+    def test_wrong_jwt_authz_payload_raises_permission_denied(self) -> None:
+        with pytest.raises(PermissionDenied):
+            self._mocked_make_authorizer({})
+
+    def test_check_deny_res(self) -> None:
+        self.authz.deny = {"test_resource": ALL}
+        with pytest.raises(PermissionDenied):
+            self.authz._check_deny()  # pylint: disable=protected-access
+        assert self.authz.outcome == DENY
+
+    def test_check_allow_all(self) -> None:
+        self.authz._check_allow_and_set_resources()  # pylint: disable=protected-access
+        assert self.authz.outcome == ALLOW
+
+    def test_check_deny_any(self) -> None:
+        self.authz.deny = {ALL: ALL}
+        with pytest.raises(PermissionDenied):
+            self.authz._check_deny()  # pylint: disable=protected-access
+        assert self.authz.outcome == DENY
+
+    def test_check_access_check_deny(self) -> None:
+        authz = self._mocked_make_authorizer({**self.token_payload, "deny": {ALL: ALL}})
+        with pytest.raises(PermissionDenied):
+            authz.check_access()
+
+    def test_check_access_outcome_deny(self) -> None:
+        authz = self._mocked_make_authorizer({**self.token_payload, "allow": {ALL: None}})
+        with pytest.raises(PermissionDenied):
+            authz.check_access()
+
+    def test__check_resource(self) -> None:
+        with patch.object(self.authz, "_deny_if_all", new_callable=MagicMock()) as mocked_deny:
+            self.authz._check_resource({"res": "xxx"})  # pylint: disable=protected-access
+            mocked_deny.assert_has_calls(
```

Comment:
This whole test is wrongly designed:
1. `_deny_if_all` is a very very simple method, it doesn't have to be mocked.
2. `_check_resource` is a protected method - shouldn't be tested.
3. By checking that one protected method triggered another protected method three times, you have frozen the implementation, not the result of the code.

Reply:
as above it this module it need to be checked.

Reply:
It depends on what you want to achieve, right now, with these tests, there is no chance to change the code without getting errors from tests. In general, that is ok but it blocks the code too much - it won't give the possibility to easily refactor the code.

By tests, we should check everything that is important meaning e.g. the result, but it does not matter how the result has been achieved 😉

Reply:
agreed to disagree offline.